feat: Do not redirect robot-users to the login portal

yabberyabber opened the pull request:
Today when you make a request using expired tokens, oidc will redirect you to a login portal. This is nice for humans who can see the portal and decide to re-auth, but this makes things hard for robot clients. The difficulty for robot clients takes two forms: first, they are unable to use the login portal that we redirect them two; second, many programs will interpret a 302 response to mean that authentication was successful.

With this change, robot clients can opt-in to the traditional 401 behavior. We have chosen to use the "X-On-Unauthorized" header to gate this behavior because it is unlikely to be used by existing clients. We preserve the existing redirect behavior when the client does not pass in an `X-On-Unauthorized` header at all. Clients can explicitly opt-in to the redirect by passing "redirect" as the value, or can opt-in to the 401 by passing "return-status-unauthorized" as the value.

**Which issue is resolved by this Pull Request:**
Relates to https://bb.infra.corp.arista.io/bug/784367

yabberyabber committed:
feat: Do not redirect robot-users to the login portal

Today when you make a request using expired tokens, oidc will
redirect you to a login portal. This is nice for humans who can
see the portal and decide to re-auth, but this makes things hard
for robot clients. The difficulty for robot clients takes two forms:
first, they are unable to use the login portal that we redirect
them two; second, many programs will interpret a 302 response to
mean that authentication was successful.

With this change, we will give these robot-clients a 401 response
which explicitly tells them that they are not authorized. Clients
should infer that they need to re-authenticate.

I've chosen the `X-Requested-With` header based on the TODO comment.
As far as I understand, the only common value for this header is
`XMLHttpRequest`, but I am just checking for a nonempty value here.

yabberyabber commented:
This is in direct conflict with Frank's comment here https://bb/bug/784367#comment-u28147193 so I need to make sure we come to some consensus before this can merge

dchenz commented on server.go:
```diff
@@ -203,7 +203,11 @@ func (s *server) authenticate(w http.ResponseWriter, r *http.Request, promptLogi
+		if requestedWith := r.Header.Get("X-Requested-With"); len(requestedWith) > 0 {
+			// Nonempty X-Requested-With header implies that this is an m2m request and
+			// so we should not try to redirect the caller to a human-centric login portal.
+			common.ReturnMessage(w, http.StatusUnauthorized, "Unauthorized")
```
There needs to be a return here so it doesn't fall through to the redirect

yabberyabber replied:
Good catch! Fixed inline.

dchenz commented:
+1

Clients that strictly need 401 can pass the header and it won't affect all clients like if we were to guess the intent using the user agent

yabberyabber committed:
Prevent fallthrough after 401 unauthorized

yabberyabber committed:
Switch to using X-On-Unauthorized header

We want to avoid using a standard header so that we don't get awkward
compatability issues with existing tools. Using this totally novel
x-on-unauthorized header will make sure we don't accentally enroll
existing clients in some behavior they weren't aware of.

This change preserves the existing default but allows clients to
optionally enroll in the new rejection behavior.

ajhfok commented:
LGTM, will wait for Danny and Frank to also give a +1 before approving

somersf commented on server.go:
```diff
@@ -202,10 +202,25 @@ func (s *server) authenticate(w http.ResponseWriter, r *http.Request, promptLogi
+		default:
+			common.ReturnMessage(w,
+				http.StatusBadRequest,
+				"Unrecognized value for X-On-Unauthorized header")
```
nit: would be nice to include the unrecognised value in the diagnostic.  With the possibility of there being several layers of wrappers, and proxies between a basic client app and this error, it would be helpful to know what the unrecognised value is, to assist in identifying where it might have come from.

yabberyabber replied:
I'm a little bit worried about the ability for bad actors to sneak stuff into your browser but I think as long as we html.ExcapeString it then I think we're safe?

somersf commented on server.go:
```diff
@@ -202,10 +202,25 @@ func (s *server) authenticate(w http.ResponseWriter, r *http.Request, promptLogi
-		// TODO: Detect "X-Requested-With" header and return 401
-		s.authCodeFlowAuthenticationRequest(w, r)
-		return nil, false
+		switch r.Header.Get("X-On-Unauthorized") {
```
nit: since the header name is used in multiple places (here, and the error message for default case), a `const xUnUnauthorized = "..."` might be nice to have

somersf reviewed (commented):
Left some comments but not feel free to ignore as you see fit.

yabberyabber committed:
Address review comments

ajhfok commented on server.go:
```diff
@@ -202,10 +202,25 @@ func (s *server) authenticate(w http.ResponseWriter, r *http.Request, promptLogi
+			logger.Infof("Failed to authenticate using authenticators. Initiating OIDC Authorization Code flow...")
+			s.authCodeFlowAuthenticationRequest(w, r)
+			return nil, false
+		case "reject":
```
nit: `return` also triggers me as a value, so I'd also vote for `return-status-unauthorized`
